File: src/main/java/tk/hiddenname/smarthome/service/task/impl/processor/EventProcessor.java
package tk.hiddenname.smarthome.service.task.impl.processor;

import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;
import tk.hiddenname.smarthome.entity.task.processing.objects.ProcessingObject;
import tk.hiddenname.smarthome.exception.NoSuchProcessorException;
import tk.hiddenname.smarthome.exception.ProcessorExistsException;
import tk.hiddenname.smarthome.exception.ProcessorNotFoundException;
import tk.hiddenname.smarthome.exception.UnsupportedObjectTypeException;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

@Component
@Scope("prototype")
@RequiredArgsConstructor
public class EventProcessor {

    private static final Logger log = LoggerFactory.getLogger(EventProcessor.class);
    private final Map<Integer, Processor> processors = new HashMap<>();

    private final ProcessorFactory processorFactory;

    public void registerProcessors(Set<ProcessingObject> objects) throws NoSuchProcessorException,
            UnsupportedObjectTypeException, ProcessorExistsException {

        for (ProcessingObject object : objects) {
            registerProcessor(object);
        }
    }

    public void registerProcessor(ProcessingObject object) throws NoSuchProcessorException,
            UnsupportedObjectTypeException, ProcessorExistsException {

        if (!processors.containsKey(object.getId())) {
            processors.put(object.getId(), processorFactory.create(object));
        } else {
            throw new ProcessorExistsException(object.getId());
        }
    }

    public void unregisterProcessors() {
        for (Integer id : processors.keySet()) {
            try {
                unregisterProcessor(id);
            } catch (ProcessorNotFoundException e) {
                log.warn(e.getMessage());
            }
        }
    }

    public void unregisterProcessor(Integer id) throws ProcessorNotFoundException {
        if (processors.containsKey(id)) {
            processors.remove(id);
        } else {
            throw new ProcessorNotFoundException(id);
        }
    }

    public void process() {
        for (Processor processor : processors.values()) {
            processor.process();
        }
    }
}

File: src/main/java/tk/hiddenname/smarthome/service/task/impl/processor/EventProcessor.kt
package tk.hiddenname.smarthome.service.task.impl.processor

import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component
import tk.hiddenname.smarthome.exception.NoSuchProcessorException
import tk.hiddenname.smarthome.exception.ProcessorExistsException
import tk.hiddenname.smarthome.exception.UnsupportedProcessingObjectTypeException
import tk.hiddenname.smarthome.exception.not_found.ProcessorNotFoundException
import tk.hiddenname.smarthome.model.task.processing.objects.ProcessingObject
import java.util.*

@Component
@Scope("prototype")
class EventProcessor(private val processorFactory: ProcessorFactory) {

    private val log = LoggerFactory.getLogger(EventProcessor::class.java)

    private val processors = HashMap<Long, Processor>()

    @Throws(NoSuchProcessorException::class, ProcessorExistsException::class,
            UnsupportedProcessingObjectTypeException::class)
    fun registerProcessors(processingObjects: Set<ProcessingObject>) {
        processingObjects.forEach {
            registerProcessor(it)
        }
    }

    @Throws(NoSuchProcessorException::class, ProcessorExistsException::class, UnsupportedProcessingObjectTypeException::class)
    fun registerProcessor(processingObject: ProcessingObject) {
        if (processors.containsKey(processingObject.id)) {
            throw ProcessorExistsException(processingObject.id)
        } else {
            processors[processingObject.id] = processorFactory.create(processingObject)
        }
    }

    fun unregisterProcessors() {
        processors.keys.forEach {
            try {
                unregisterProcessor(it)
            } catch (e: ProcessorNotFoundException) {
                log.warn(e.message)
            }
        }
    }

    @Throws(ProcessorNotFoundException::class)
    fun unregisterProcessor(id: Long) {
        if (processors.containsKey(id)) {
            processors.remove(id)
        } else {
            throw ProcessorNotFoundException(id)
        }
    }

    fun process() {
        processors.values.forEach {
            it.process()
        }
    }
}